src/python/pants/engine/internals/build_files.py

```diff
@@ -387,16 +389,91 @@ def _extract_env_vars(
     )
 
     # Process synthetic targets.
-    for address_map in address_maps:
-        for synthetic in synthetic_address_maps:
-            synthetic.process_declared_targets(address_map)
-            synthetic.apply_defaults(frozen_defaults)
+
+    def apply_defaults(tgt: TargetAdaptor) -> TargetAdaptor:
+        default_values = frozen_defaults.get(tgt.type_alias)
+        if default_values is None:
+            return tgt
+        return tgt.with_new_kwargs(**default_values, **tgt.kwargs)
+
+    name_to_path_and_synthetic_target: dict[str, tuple[str, TargetAdaptor]] = {}
+    for synthetic_address_map in synthetic_address_maps:
+        for name, target in synthetic_address_map.name_to_target_adaptor.items():
+            name_to_path_and_synthetic_target[name] = (
+                synthetic_address_map.path,
+                apply_defaults(target),
+            )
+
+    name_to_path_and_declared_target: dict[str, tuple[str, TargetAdaptor]] = {}
+    for declared_address_map in declared_address_maps:
+        for name, target in declared_address_map.name_to_target_adaptor.items():
+            name_to_path_and_declared_target[name] = (declared_address_map.path, target)
+
+    # We copy the dict so we can modify the original in the loop.
+    for name, (
+        declared_target_path,
+        declared_target,
+    ) in name_to_path_and_declared_target.copy().items():
+        declared_target_kwargs = dict(declared_target.kwargs)
+        extend_synthetic = declared_target_kwargs.pop("_extend_synthetic", False)
```

Comment:
have you checked the performance on a "large"-ish test repo for this, compared with the previous code path?

I worry that with many targets, this unconditional extra overhead could be costly? (no facts, just a hunch.)

Reply:
Well, asymptotically this is faster than the previous code because that was O(m*n) on the number of synthetic and declared targets in a single family. But I see your point that this involves creating new kwargs for every single target even though almost none of them need it. Will fix that.

Reply:
I note that in the unlikely case where someone set `_extend_synthetic` explicitly, but to False, we still have to remove that kwarg from the target, as nothing downstream will know what to do with it.